### PR TITLE
Fix duplicated volume issue

### PR DESCRIPTION
**Pull request by wilsonwang371**
## Why are these changes needed?

some corner cases can be triggered and the error below is returned

```
Invalid value: \"/dev/shm\": must be unique", 
```

## Related issue number

#688

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [X] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/common/pod.go [hunk 1]
```diff
@@ -711,6 +711,14 @@ func addEmptyDir(container *v1.Container, pod *v1.Pod, volumeName string, volume
 	if checkIfVolumeMounted(container, pod, volumeMountPath) {
 		return
 	}
+
+	// 0) Skip adding the volume if it is already in the list of volumes
+	for _, volume := range container.VolumeMounts {
+		if volume.Name == volumeName && volume.MountPath == volumeMountPath {
+			return
+		}
+	}
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/common/pod.go:720]: Thanks for fixing this!

Would you mind factoring this out into a function `checkIfVolumeMountExists` similar to the `checkIfVolumeExists` function just below it?

Comment (wilsonwang371) [reply ray-operator/controllers/ray/common/pod.go:720]: sure. let me first test it with @iycheng and see if this resolves the issue.

**Review by Jeffwan — APPROVED** [review]

File: ray-operator/controllers/ray/common/pod.go (hunk 1 above)
Comment (Jeffwan) [reply ray-operator/controllers/ray/common/pod.go:720]: Agree. This should be easily verified in local env.

File: ray-operator/controllers/ray/common/pod.go [hunk 2]
```diff
@@ -711,6 +711,14 @@ func addEmptyDir(container *v1.Container, pod *v1.Pod, volumeName string, volume
 	if checkIfVolumeMounted(container, pod, volumeMountPath) {
 		return
 	}
+
+	// 0) Skip adding the volume if it is already in the list of volumes
+	for _, volume := range container.VolumeMounts {
+		if volume.Name == volumeName && volume.MountPath == volumeMountPath {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:717]: I have some concerns about this line. MountPath needs to be unique. Hence, if users specify a volume with a unique name but the same MountPath (`/tmp/ray` or `/dev/shm`), Kubernetes apiserver will report error. 

Take the following YAML file as an example, when you use `kubectl apply` to apply the YAML file, Kubernetes apiserver will report `The Pod "test-pd" is invalid: spec.containers[0].volumeMounts[1].mountPath: Invalid value: "/cache": must be unique`.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pd
spec:
  containers:
  - image: ubuntu:latest
    command:
    - "sleep"
    - "604800"
    name: test-container
    volumeMounts:
    - mountPath: /cache
      name: cache-volume
    - mountPath: /cache
      name: cache-volume-2
  volumes:
  - name: cache-volume
    emptyDir: {}
  - name: cache-volume-2
    emptyDir: {}
```

**Review by kevin85421 — CHANGES_REQUESTED** [review]

File: ray-operator/controllers/ray/common/pod.go (hunk 2 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/common/pod.go:717]: In my opinion, the logic should be

```go
	// 0) Skip adding the volume if it is already in the list of volumes
	for _, volume := range container.VolumeMounts {
		if volume.MountPath == volumeMountPath {
                     return
                }
        }
```

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:717]: This is a good point. To be make the check even stricter, we can avoid duplicating either name or path. Also, the variable iteration variable should be `volumeMount`.

I suggest
```go
	// 0) Skip adding the volume if it is already in the list of volumes
	for _, volumeMount := range container.VolumeMounts {
		if volumeMount.Name == volumeName || volumeMount.MountPath == volumeMountPath {
			return
		}
	}
```

Comment (kevin85421) [reply ray-operator/controllers/ray/common/pod.go:717]: (1) Duplicate `volumeMount.Name` is allowed in Kubernetes. That is, two MountPaths are the same directory. This is weird, but files written to both MountPaths can be persisted. 

(2) If we skip the duplicate `volumeMount.Name`, it is possible to have no `volumeMount` for `/tmp/ray` or `/dev/shm`. Hence, some important data may not be persisted.


Both (1) and (2) may cause users confusing. For me, I will report an error if any duplicate `volumeMount.Name` or `volumeMount.MountPath` exists, but this may be too aggressive.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:717]: Makes sense, de-duping mount path seems enough.

Comment (wilsonwang371) [reply ray-operator/controllers/ray/common/pod.go]: updated

**Review by DmitriGekhtman — APPROVED** [review]
Unit test appreciated but not required.

**Review by kevin85421 — COMMENTED** [review]
Thank @wilsonwang371 for the update! Can you rebase the branch? I have added some config tests in the latest commit. In addition, unit tests appreciated. Thank you!

**Review by DmitriGekhtman — CHANGES_REQUESTED** [review]
Mm, actually, I do want to request a unit test -- testing is pretty important to the stability of this codebase.

File: ray-operator/controllers/ray/common/pod.go [hunk 3]
```diff
@@ -755,12 +757,7 @@ func makeEmptyDirVolume(container *v1.Container, volumeName string, storageMediu
 func checkIfVolumeMounted(container *v1.Container, pod *v1.Pod, volumeMountPath string) bool {
 	for _, mountedVol := range container.VolumeMounts {
 		if mountedVol.MountPath == volumeMountPath {
-			for _, podVolume := range pod.Spec.Volumes {
```
Comment (wilsonwang371) [inline ray-operator/controllers/ray/common/pod.go:758]: @DmitriGekhtman Dima, what is the original reason for adding this pod.Spec.Volumes comparison? I discussed with @scarlet25151 about this and we are not sure whether we need this part.

Comment (wilsonwang371) [reply ray-operator/controllers/ray/common/pod.go:758]: It is the first commit. And you are the one that made changes to code at the closest place.... @DmitriGekhtman 

image

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:758]: Oh, I just reused the existing code to add a new volume mount for autoscaler logs.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:758]: I don't think we need the extra check. Indeed, it's incorrect, as it allows redundant mountPaths!
Thanks for fixing it!

Comment (wilsonwang371) [reply ray-operator/controllers/ray/common/pod.go:758]: thanks for the info

Comment (wilsonwang371) [reply ray-operator/controllers/ray/common/pod.go:758]: FYI: @scarlet25151

**Comment by DmitriGekhtman** [comment]
@wilsonwang371 this will be good to merge after a basic unit test is added

**Review by DmitriGekhtman — APPROVED** [review]

**Review by kevin85421 — APPROVED** [review]

**Comment by kevin85421** [comment]
@wilsonwang371 is it ready to merge? The PR title is "WIP". Thank you!

**Comment by DmitriGekhtman** [comment]
Deleted the WIP :) . Merging.

**Comment by wilsonwang371** [comment]
> Deleted the WIP :) . Merging.

this is really fast..... i was thinking about a few more tests... although it is fine to merge now...

**Comment by DmitriGekhtman** [comment]
My bad! Sometimes I'm a little trigger happy.
Feel free to add tests in a follow-up!
